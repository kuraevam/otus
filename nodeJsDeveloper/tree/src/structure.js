import fs from 'node:fs';
import path from 'node:path';
import renderStructure from './utils/renderStructure.js';

function getStructureDirectory(folderPath, depth) {
  const structure = {
    name: path.basename(folderPath),
  };

  if (depth === 0) {
    return structure;
  }
  depth -= 1;

  const childPaths = fs.readdirSync(folderPath).map((fileName) => path.join(folderPath, fileName));

  const items = [];
  for (const childPath of childPaths) {
    if (fs.lstatSync(childPath).isFile()) {
      items.push({
        name: path.basename(childPath),
      });
    }

    if (fs.lstatSync(childPath).isDirectory()) {
      const structureChild = getStructureDirectory(childPath, depth);
      items.push(structureChild);
    }
  }

  if (items.length) {
    structure.items = items;
  }

  return structure;
}

function getJsonFromFile(filePath) {
  if (!fs.existsSync(filePath) || !fs.statSync(filePath).isFile()) {
    throw Error('path does not exist');
  }

  return JSON.parse(fs.readFileSync(filePath, 'utf8'));
}

function getStructureJson(json, depth) {
  const structure = {
    name: json.name,
  };

  if (depth === 0 || !Array.isArray(json.items)) {
    return structure;
  }
  depth -= 1;

  structure.items = [];
  for (const item of json.items) {
    const structureChild = getStructureJson(item, depth);
    structure.items.push(structureChild);
  }

  return structure;
}

export function showDirectoryFileStructure(folderPath, depth) {
  const structure = getStructureDirectory(folderPath, depth);
  renderStructure(structure);
}

export function showJsonStructure(folderPath, depth) {
  const json = getJsonFromFile(folderPath);
  const structure = getStructureJson(json, depth);
  renderStructure(structure);
}
