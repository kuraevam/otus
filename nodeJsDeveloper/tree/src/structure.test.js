import path from 'node:path';
import { getTreeByDirectory, getTreeByJson } from './structure.js';

describe('getTreeByJson', () => {
  test('it should return the tree', async () => {
    const folderPath = path.join(__dirname, '..', './resources/index.json');
    const result = getTreeByJson(folderPath, 4);
    expect(result).toMatchSnapshot();
  });
});

describe('getTreeByDirectory', () => {
  test('it should return the tree', async () => {
    const folderPath = path.join(__dirname, '..', './resources/tree');
    const result = getTreeByDirectory(folderPath, 4);
    expect(result).toMatchSnapshot();
  });
});
