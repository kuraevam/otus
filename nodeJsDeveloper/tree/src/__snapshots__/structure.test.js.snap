// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTreeByDirectory it should return the tree 1`] = `
"tree
|------file1
|------file2
|------folder1
|   |-----file3
|   |-----file4
|   |-----folder2
|   |  |-----file5
|   |  |-----file6"
`;

exports[`getTreeByJson it should return the tree 1`] = `
"1
|------2
|   |-----3
|   |-----4
|------5
|   |-----6
|   |  |-----7
|   |  |  |-----8"
`;
